extensions/notebook/src/book/remoteBookController.ts

```diff
@@ -0,0 +1,290 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as request from 'request';
+import * as fs from 'fs';
+import * as nls from 'vscode-nls';
+import * as vscode from 'vscode';
+import * as path from 'path';
+import * as zip from 'adm-zip';
+import * as tar from 'tar';
+import * as utils from '../common/utils';
+import { RemoteBookDialogModel } from '../dialog/remoteBookDialog';
+
+const localize = nls.loadMessageBundle();
+const msgRemoteBookDownloadProgress = localize('msgRemoteBookDownloadProgress', "Remote Book download is in progress");
+const msgRemoteBookDownloadComplete = localize('msgRemoteBookDownloadComplete', "Remote Book download is complete");
+const msgRemoteBookDownloadError = localize('msgRemoteBookDownloadError', "Error while downloading remote Book");
+const msgRemoteBookUnpackingError = localize('msgRemoteBookUnpackingError', "Error while decompressing remote Book");
+const msgRemoteBookDirectoryError = localize('msgRemoteBookDirectoryError', "Error while creating remote book directory");
+const msgTaskName = localize('msgTaskName', "Downloading Remote Book");
+
+const msgResourceNotFound = localize('msgResourceNotFound', 'Resource not Found');
+const msgBookNotFound = localize('msgBookNotFound', 'Books not Found');
+const msgReleaseNotFound = localize('msgReleaseNotFound', 'Releases not Found');
+
+export class RemoteBookController {
+	private _book: RemoteBook;
+	private readonly assetNameRE = /([a-zA-Z0-9]+)(?:-|_)([a-zA-Z0-9.]+)(?:-|_)([a-zA-Z0-9]+).(zip|tar.gz|tgz)/;
+	constructor(public model: RemoteBookDialogModel) {
+	}
+
+	public async setRemoteBook(url: URL, remoteLocation: string, asset?: IAssets): Promise<void> {
+		if (remoteLocation === 'GitHub') {
+			this._book = new GitHubRemoteBook(url, asset.browser_download_url);
+		} else {
+			this._book = new SharedRemoteBook(url);
+		}
+		return await this._book.createLocalCopy();
+	}
+
+	public async fetchGithubReleases(url: URL): Promise<IReleases[]> {
+		let options = {
+			headers: {
+				'User-Agent': 'request'
+			}
+		};
+		return new Promise<IReleases[]>((resolve, reject) => {
+			request.get(url.href, options, (error, response, body) => {
+				if (error) {
+					return reject(error);
+				}
+
+				if (response.statusCode === 404) {
+					return reject(msgResourceNotFound);
+				}
+
+				if (response.statusCode !== 200) {
+					return reject(response.statusCode);
+				}
+				let releases = JSON.parse(body);
+				let bookReleases: IReleases[] = [];
+				if (releases) {
+					let keys = Object.keys(releases);
+					keys = keys.filter(key => {
```

Comment:
Why are you using filter here? forEach seems like a better choice for what you're doing.